lib/internal/async_hooks.js

```diff
@@ -104,12 +108,24 @@ const emitPromiseResolveNative =
     emitHookFactory(promise_resolve_symbol, 'emitPromiseResolveNative');
 
 const topLevelResource = {};
+// Contains a [stack height] -> [resource] map for things pushed onto the
+// async resource stack from JS.
+const jsResourceStack = new SafeMap();
+// Contains either a single key (null) or nothing. If the key is present,
+// this points to the current async resource.
+const cachedResourceHolder = new SafeMap();
```

Comment:
Can this single key Map be replaced with a variable slot that can be either null or the resource object?

Reply:
The problem is that it needs to be some sort of object that can be easily modified from C++ as well (at least until we have weak references in JS without a flag)… I picked a map because clearing it from C++ never throws exceptions, but it could also be a single-entry Array or object.

Reply:
Alternatively, yes, this could be a single variable slot, but that might become an issue with memory retention in some weird edge cases.

Reply:
Why not just use a symbol property? 🤔

Reply:
@qard Symbol property on what/in place of what? If you’re talking about the map key, that would work just as well, yes. If you’re talking about replacing the map with an object with a single symbol key, that’s harder to clear from C++ than a map, I’d say?

Reply:
Btw, we can remove quite a bit of complexity here once we have weak refs without a flag. That would be the actual correct thing to use here. (The internal `WeakRef` we use elsewhere most likely isn’t fast enough, so we’ll need to wait for V8’s one.)